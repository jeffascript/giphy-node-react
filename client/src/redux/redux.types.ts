import { SerializedError } from '@reduxjs/toolkit';
import { Status } from '../ContextAPI/types.context';

export type Bookmark = {
    id: string;
    url: string;
};

export interface IBookmarkState {
    status: Status.Idle | Status.Resolved | Status.Rejected | Status.Pending;
    LoadedCount: number;
    bookmarks: Bookmark[];
    bookmarkError?: string | null | SerializedError;
}
