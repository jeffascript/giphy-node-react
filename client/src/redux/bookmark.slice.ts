import { createSlice } from '@reduxjs/toolkit';
import { Bookmark, IBookmarkState } from './redux.types';
import { Status } from '../ContextAPI/types.context';
/* eslint-disable @typescript-eslint/no-use-before-define */
/* eslint-disable no-param-reassign */

import { AppDispatch, RootState } from './store';

export const addBookmark = ({ url, id }: Bookmark) => {
    return async (dispatch: AppDispatch, getState: () => RootState) => {
        dispatch(bookmarkPending());

        const rootState = getState() as RootState;
        const found = rootState.allBookmarkState.bookmarks.some((el) => el.id === id);
        if (!found) {
            dispatch(addBookmarkSuccess({ url, id }));
        } else {
            dispatch(addBookmarkFailure('GiF Already Bookmarked'));
        }
    };
};

export const removeBookmark = ({ url, id }: Bookmark) => {
    return (dispatch: AppDispatch, getState: () => RootState) => {
        dispatch(bookmarkPending());
        const rootState = getState() as RootState;
        const found = rootState.allBookmarkState.bookmarks.some((el) => el.id === id);
        if (found) {
            dispatch(removeBookmarkSuccess({ url, id }));
        }
    };
};

const initialState = {
    LoadedCount: 0,
    bookmarks: [],
    bookmarkError: null,
    status: Status.Idle,
} as IBookmarkState;

// new Redux toolkit uses Immer, hence can be mutated  : cf: https://redux-toolkit.js.org/usage/immer-reducers
const bookmarkHandler = createSlice({
    name: 'allBookmarkState',
    initialState,
    reducers: {
        bookmarkPending: (state) => {
            state.status = Status.Pending;
            state.bookmarkError = null;
        },

        addBookmarkSuccess: (state, action) => {
            state.bookmarks.push(action.payload);
            state.LoadedCount += 1;
            state.status = Status.Resolved;
        },
        addBookmarkFailure: (state, action) => {
            if (action.payload) state.bookmarkError = action.payload;
            state.status = Status.Rejected;
        },

        removeBookmarkSuccess: (state, action) => {
            state.LoadedCount -= 1;
            state.status = Status.Resolved;
            state.bookmarks = state.bookmarks.filter((item) => item.id !== action.payload.id);
        },

        clearAllBookmarkData: (state) => {
            return {
                ...state,
                LoadedCount: 0,
                bookmarks: [],
                bookmarkError: null,
                status: Status.Idle,
            };
        },
    },
    //   extraReducers: () => {},
});

export const { addBookmarkFailure, addBookmarkSuccess, removeBookmarkSuccess, bookmarkPending, clearAllBookmarkData } =
    bookmarkHandler.actions;

export default bookmarkHandler.reducer;
