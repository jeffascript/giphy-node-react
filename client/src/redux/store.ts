import { configureStore } from '@reduxjs/toolkit';
import { persistReducer } from 'redux-persist';
import { combineReducers } from 'redux';
import storage from 'redux-persist/lib/storage';
import bookmarkHandlerReducer from './bookmark.slice';

const reducers = combineReducers({
    allBookmarkState: bookmarkHandlerReducer,
});

const persistConfig = {
    key: 'root',
    storage,
};

const persistedReducer = persistReducer(persistConfig, reducers);

const store = configureStore({
    reducer: persistedReducer,
    middleware: (getDefaultMiddleware) =>
        getDefaultMiddleware({
            serializableCheck: false,
        }),
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export default store;
