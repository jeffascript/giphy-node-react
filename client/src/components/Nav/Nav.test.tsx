import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from '@testing-library/react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import ServiceListingScreen from './GalleryIndicator';
import serviceReducer from '../../redux/bookmark.slice';

const store = createStore(serviceReducer);
configure({ adapter: new Adapter() });

const ReduxProvider = ({ children, reduxStore }) => <Provider store={reduxStore}>{children}</Provider>;

describe('Screen/ServiceListingScreen', () => {
    it('renders correctly ', () => {
        const wrapper = shallow(
            <Provider store={store}>
                <ServiceListingScreen />
            </Provider>,
        );
        // const tree = renderer(wrapper).toJSON();
        // expect(tree).toMatchSnapshot();
    });
});
