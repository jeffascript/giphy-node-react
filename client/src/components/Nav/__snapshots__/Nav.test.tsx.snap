// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the NavBar with Indicator 1`] = `
<DocumentFragment>
  <div
    class="css-10g4ob8"
  >
    <div
      class="css-1kvivmg"
    >
      <h2>
        Search For Your GIFS!!! 
      </h2>
      <div
        class="css-iavuzk"
      >
        <div
          class="rs-badge none rs-badge-wrapper"
        >
          <div>
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="font-size: 40px;"
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 512V48C0 21.49 21.49 0 48 0h288c26.51 0 48 21.49 48 48v464L192 400 0 512z"
              />
            </svg>
          </div>
          <div
            class="rs-badge-content"
          >
            0
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
</DocumentFragment>
`;
